lang/syntax/attribute.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2025 The µcad authors <info@ucad.xyz>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! Attribute syntax entities.
+
+use crate::{src_ref::*, syntax::*};
+
+/// An attribute item.
+#[derive(Debug, Clone)]
+pub enum Attribute {
+    /// A tag attribute `foo::bar`
+    Tag(QualifiedName),
+    /// A call attribute `foo::bar(baz = 32)`
+    Call(Call),
+    /// A name value attribute `foo::baz = baz`
+    NameValue(QualifiedName, Expression),
+}
+
+impl Attribute {
+    /// Return qualified name for this attribute.
+    pub fn qualified_name(&self) -> &QualifiedName {
+        match self {
+            Attribute::Tag(qualified_name) => qualified_name,
+            Attribute::Call(call) => &call.name,
+            Attribute::NameValue(qualified_name, _) => qualified_name,
+        }
+    }
+}
+
+impl std::fmt::Display for Attribute {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Attribute::Tag(qualified_name) => writeln!(f, "#[{qualified_name}]"),
+            Attribute::Call(call) => writeln!(f, "#[{call}]"),
+            Attribute::NameValue(qualified_name, expression) => {
+                writeln!(f, "#[{qualified_name} = {expression}]")
+            }
+        }
+    }
+}
+
+impl SrcReferrer for Attribute {
+    fn src_ref(&self) -> crate::src_ref::SrcRef {
+        match self {
+            Attribute::Tag(qualified_name) => qualified_name.src_ref(),
+            Attribute::Call(call) => call.src_ref(),
+            Attribute::NameValue(qualified_name, _expression) => {
+                // FIXME: This should return a merged src_ref of `qualified_name` and `expression`
+                qualified_name.src_ref()
```

Comment:
use `SrcRef::merge()`:

```suggestion
                SrcRef::merge(qualified_name,_expression)
```